README.md

```diff
@@ -1 +1,257 @@
+
+Data purporting to conform to an IPLD encoding (such as [DAG-JSON](https://ipld.io/specs/codecs/dag-json/spec/)) MUST be validated prior to signature verification. This can be as simple as round-trip decoding/encoding the JSON and checking that the hash matches. A validation error MUST be signalled if it does not match. 
+
+> [Implementers] may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
+>
+> — [DAG-JSON Spec](https://ipld.io/specs/codecs/dag-json/spec/)
+
+As it is critical for guarding against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+### 2.1.1 Caching & Invalidation
+
+Signing CIDs has two additional caching effects
+
+* Signing CIDs enables a simple strategy for caching validation by CID
+* Such a strategy also requires accounting for revocation of the signing keys, and so need to be tagged with this additional metadata
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `0x55`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+After being decoded from [LEB128](https://en.wikipedia.org/wiki/LEB128)s, a varsig includes the following segments:
+
+```abnf
+varsig = %x34 varsig-content-encoding varsig-header varsig-body
+varsig-content-encoding = LEB128 ; The multicodec describing what's being signed over
+varsig-header = LEB128 ; Usually the public key code from Multicodec
+varsig-body = *LEB128; Zero or more segments required by the kind of varsig (e.g. raw bytes, hash algorithm, etc)
+```
+
+For example, here is a EdDSA signature for content encoded as DAG-PB:
+
+`0x3470edae3784f03f9ee1163382fa6efa73b0c31ecf58c899c836709303ba4621d1e6df20e09aaa568914290b7ea124f5b38e70b9b69c7de0d216880eac885edd41c302`
+
+### 3.1 Varsig Prefix
+
+The varsig prefix MUST be `0x34`.
+
+### 3.2 Content Multicodec Prefix
+
+The [multicodec](https://github.com/multiformats/multicodec) prefix of the encoding scheme used before it's signed over. The default encoding SHOULD be [`0x55` "raw binary"](https://github.com/multiformats/multicodec/blob/master/table.csv#L40).
+
+Many signature schemes depend on a hash function. Algorithm-sensitive hashing MUST be captured in the Varsig [header algorithm](#3-1-3-signature-header) or [body](#3-1-4-varsig-body), and so MUST NOT be captured in the Content Multicodec Prefix field.
+
+Some examples include:
+
+* `0x55` for raw bytes (no special encoding)
+* `0x70` for DAG-PB
+* `0x0129` for DAG-JSON
+* `0x0202` for [CAR](https://ipld.io/specs/transport/car/) files
```

Comment:
I'm still pretty sure that we want this, but could be convinced that we only sign over raw bytes if this is controversial. The argument from before was that we don't want to leave it up to interpretation, and so should be explicit.